evm2wast.c

```diff
@@ -0,0 +1,320 @@
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <dirent.h>
+#include <math.h>
+#include <errno.h>
+#include <stdarg.h>
+#include "evm2wast.h"
+#include "gadgets.h"
+int digits(unsigned long x) { return (floor(log10(abs(x))) + 1); }
+static size_t stack_delta;
+static size_t stack_high;
+static size_t stack_low;
+static size_t gas_count;
+//static char *callbacks[12];
+//static char *deps[256];
+
+static char *build_jump_map(size_t *segments, size_t len)
+{
+	size_t i;
+	char *brtable = calloc(len*(2+digits(len)), 1);
+	char *wasm = calloc(pow((67+2*(digits(len))), len), 1);
+	if(brtable == 0 || wasm == 0) goto err;
+	for(i=0; i<len; i++)
+	{
+		snprintf(brtable + strlen(brtable), 2+digits(len)+1, " $%ld", i);
+		snprintf(wasm + strlen(wasm), 68+(2*(digits(len))),
+		"(if(i32.eq(get_local $jump_dest)(i32.const $%ld))(then (br $%ld))(else %s))",
+		segments[i], i, wasm);
+	}
+	size_t s = strlen(wasm) + 358;
+	wasm = realloc(wasm, s);
+	if(wasm == 0) goto err;
+	snprintf(wasm + strlen(wasm), s,
+	"(block $0 (if (i32.eqz (get_global $init)) (then (set_global $init (i32.const 1)) (br $0)) (else (if (i32.eq (get_global $cb_dest) (i32.const 0)) (then %s) (else $cb_dest get_global $cb_dest (set_global $cb_dest (i32.const 0)) (br_table $0 ${brTable}))))))", wasm);
+	return wasm;
+err:
+	free(brtable);
+	free(wasm);
```

Comment:
This could result in `free`ing `NULL` in the case of `brtable` or `wasm` allocation fails.